### ISSUE TITLE
Video unavailable after 1 hour

### ISSUE THREAD

**Issue by tescophil** [tchellomello/python-ring-doorbell#191]
Video from the camera (preview image and the stream) is not available one hour after the last event, or one hour after start-up.

After a new motion or ding event the stored video for the event will be available for 1 hour then go offline.

If HA is restarted then the last video stored will again be available for one hour before it goes offline.

No errors in the HA log. (I assume there is some token involved that is expiring after the hour ?)

**Comment by Tosko4** [comment]
Hey, I wanted to raise your awareness, many more have this issue. Home Assistant (balloob) suggested we start here to try to solve this issue. Please @tchellomello , have a look at the following github issue:

https://github.com/home-assistant/home-assistant/issues/30927#issuecomment-575946513

**Comment by dshokouhi** [comment]
Can you guys check one thing.  When this issue occurs, go outside and trigger a recording for 10 seconds.  Wait a few minutes and check HA again.  I think what is happening is Ring is expiring the last video URL that we use to display the video.  To check this when the issue happens note the `video_url` attribute for the camera in HA and then create the event.  Wait about 3-5 minutes for the video to generate and then check if the attribute updated.

I just did this myself when I saw the issue on one of my cameras and as soon as the next video was created it worked.   If that is the case then its probably part of the auth cleanup Ring did and most likely intentional since you can see in the URL that its set to expire in 1 hour.  But check this out on your end and see if you get the same results.

**Comment by thehaxxa** [comment]
> Can you guys check one thing. When this issue occurs, go outside and trigger a recording for 10 seconds. Wait a few minutes and check HA again. I think what is happening is Ring is expiring the last video URL that we use to display the video. To check this when the issue happens note the `video_url` attribute for the camera in HA and then create the event. Wait about 3-5 minutes for the video to generate and then check if the attribute updated.
> 
> I just did this myself when I saw the issue on one of my cameras and as soon as the next video was created it worked. If that is the case then its probably part of the auth cleanup Ring did and most likely intentional since you can see in the URL that its set to expire in 1 hour. But check this out on your end and see if you get the same results.

Confirmed this is the Case on my end.

It seems the URL expires after 3600 seconds. 
Given this is the case, the API is working correctly and this is unfortunately a server side change.

_As an alternative solution:_
I would prefer if Home Assistant caches the last available video and runs a check every '_x_' minutes to download new captured motion. This way the last motion will always be available. When the stream is opened it will check if a new video exists otherwise if it returns 403/404 etc. it just plays the cached file. 

Any thoughts on this method? I thought I would seek opinions, before I open an issue on Home Assistant's Issues Page.

It also means it won't be delayed in Home Assistant with a slow server response. Which is very apparent often taking several seconds in Lovelace before the component's video is loaded in HASS.

**Comment by dshokouhi** [comment]
You can do this today actually: https://www.home-assistant.io/integrations/ring/#saving-the-videos-captured-by-your-ring-door-bell

HA just asks the parent library for the video and displays it, it does not cache it for any integration AFAIK.  I think the current python script should satisfy this need right?

**Comment by thehaxxa** [comment]
> You can do this today actually: https://www.home-assistant.io/integrations/ring/#saving-the-videos-captured-by-your-ring-door-bell
> 
> HA just asks the parent library for the video and displays it, it does not cache it for any integration AFAIK. I think the current python script should satisfy this need right?

Sure its doable, calling a Python Script to download the latest video file, overwriting last file as stated.

Setup something like this: https://community.home-assistant.io/t/ffmpeg-camera-from-local-video-file/138843/5 to play the local file and configure as before. 

Ideally though this would be integrated into the component or provided as an option as the setup is fairly involved, and I imagine most people would like the behavior consistent to what it was before.

**Comment by steve-gombos** [comment]
I've been looking into this today, but I actually think this is an issue with HA itself.  I ran two tests earlier letting it go for an hour after a restart and was able to verify that the camera video/image wouldn't load.  

I have forked HA and made a couple changes (https://github.com/steve-gombos/home-assistant/tree/ring-camera-fix) and have been running for about a couple hours and with the changes I made I am still able to see the last event in the camera beyond an hour now.  I am just verifying a couple of other things and letting it run to see if there are any other issues. I think the issue is a bad utc time check in the camera entity, so it would do an early return instead of getting a new video_url.  Feel free to pull down my fork and let me know what kind of results you see.

I did notice that when I trigger motion at my doorbell the image doesn't update on the camera right away, not sure the cause of that potentially, but the camera entity does eventually update with the right event video/image.
